### PR TITLE
Add Description and Default fields to TaskParams

### LINKED ISSUE #190: Expand Task Param fields
## Expected Behavior
Tasks should be reusable. When you look at a Task, it should be clear how to use it, specifically how to provide the right parameters and what they are. There should also be defaults for parameters, i.e. each Param should look like this:

```yaml
  - name: DOCKERFILE_NAME
    description: The name of the Dockerfile
    default: Dockerfile
```

* Description should be required
* Default should be optional

_Please also update examples._

## Actual Behavior
Currently Task Params have only one field: `name`.

## Additional Info
See [knative build template docs](https://github.com/knative/docs/blob/master/build/build-templates.md#example-template).

Question: perhaps we want this functionality for resources as well?

### PR DESCRIPTION
**Pull request by tanner-bruce**
This commit adds the `Description` and `Default` fields to the
TaskParams object. Description is a mandatory field to be used for
describing what the Param is. Default is optional, and may be used to
specify a default value for the param.

Adding these fields will make the pipeline easier to understand.

Also fixes a bug in `resources.AddInputResources` where if the Task has
Inputs defined, AddInputResources assumes there *must* be one of type
GitResource, which is not necessarily the case.

Fixes #190

### CODE REVIEW

**Comment by tanner-bruce** [comment]
Whoops, used a different email as the author.

**Comment by tejal29** [comment]
i dont feel like `Description` should be mandatory.  Users can create Tasks and  add them to their cluster. These tasks could be shared across clusters or can be private and used as POC.
Making them mandatory would just add another friction point especially if its just for testing.
Ppl can just add empty string or non relevant blobs to get around it.
I the validation should not dictate good practices and leave it up to the user.

WDYT?

**Comment by shashwathi** [comment]
@tejal29 I agree with that. `description` provides a place to add explanation but making it mandatory seems like enforcing opinionated design. Enforcing will also mean more number of things a user has to declare to get pipeline up and running. I vote for making `description` optional.

**Comment by tanner-bruce** [comment]
I agree with @shashwathi and @tejal29 on making description not mandatory. @bobcatfish if we are all in agreement I will remove that part.

File: pkg/apis/pipeline/v1alpha1/taskrun_types.go [hunk 1]
```diff
@@ -94,15 +94,15 @@ type TaskRunStatus struct {
...
-func (ts *TaskRunStatus) InitializeConditions() {
-	taskRunCondSet.Manage(ts).InitializeConditions()
+func (tr *TaskRunStatus) InitializeConditions() {
+	taskRunCondSet.Manage(tr).InitializeConditions()
 }
 
 // SetCondition sets the condition, unsetting previous conditions with the same
 // type as necessary.
-func (ts *TaskRunStatus) SetCondition(newCond *duckv1alpha1.Condition) {
+func (tr *TaskRunStatus) SetCondition(newCond *duckv1alpha1.Condition) {
```
Comment (bobcatfish) [inline pkg/apis/pipeline/v1alpha1/taskrun_types.go]: thanks for cleaning these up!

**Comment by bobcatfish** [comment]
> i dont feel like Description should be mandatory. Users can create Tasks and add them to their cluster. These tasks could be shared across clusters or can be private and used as POC.

@tejal29 I think there might be some confusion - this is only in the Task definitions. Having mandatory descriptions would actually make it easier to reuse the Tasks, b/c there would be a description of each parameter.

Users of the Tasks do not need to fill in any kind of Description field, only authors of Tasks, e.g. in one of the examples @tanner-bruce updated:

```yaml
apiVersion: pipeline.knative.dev/v1alpha1
kind: Task
metadata:
  name: build-push
  namespace: default
spec:
    inputs:
        resources:
            - name: workspace
              type: git
        params:
            - name: pathToDockerFile
              description: The path to the dockerfile to build
    outputs:
        resources:
            - name: builtImage
              type: image
    buildSpec:
          steps:
          - name: build-and-push
            image: gcr.io/kaniko-project/executor
            args:
            - --dockerfile=${pathToDockerFile}
            - --destination=$builtImage
```

The author of this Task would add a description to the param, so that the potentially many users of this Task would understand what that param is for.

> I feel the validation should not dictate good practices and we shd leave it up to the user and their processes(e.g. code review etc) to determine it.
> description provides a place to add explanation but making it mandatory seems like enforcing opinionated design.

As an opinionated person myself, I feel pretty strongly opinionated that if you put a parameter somewhere, you should include a description of it.

But I see what you're saying, that's why we leave this kind of thing to linters I guess. If it was up to me I'd make this field mandatory but I'm okay with it being optional.

**Comment by tejal29** [comment]
> @tejal29 I think there might be some confusion - this is only in the Task definitions. Having mandatory descriptions would actually make it easier to reuse the Tasks, b/c there would be a description of each parameter.

Yes, i am aware this is only for task description and task-runs dont' have it  nor the pipelines which user will end up writing the most. But again to my point, if i am writing a test task, i would end up filling in "foo" or "dummy" similar to just get past the validation since `Description` does not actually matter.

I am strong advisor where "Tools encourage good practices (in this case we are - by adding the description field) but let users enforce them"

File: examples/build_task.yaml [hunk 2]
```diff
@@ -10,6 +10,7 @@ spec:
               type: git
         params:
             - name: pathToDockerFile
+              description: The path to the dockerfile to build
```
Comment (pivotal-nader-ziada) [inline examples/build_task.yaml]: a good candidate to put an example of default value here

**Comment by pivotal-nader-ziada** [comment]
I agree with @tejal29 that enforcing a description will make users just put some not useful text, so I also prefer making it optional

**Pull request merged**